File: modules/m2y-crypto/src/M2YCryptoModule.ts
import { requireNativeModule } from 'expo';

import type { M2YCryptoNativeModule } from './M2YCrypto.types';

const nativeModule = requireNativeModule<M2YCryptoNativeModule>('M2YCrypto');

export function getSpikeInfo(): unknown {
  return nativeModule.getSpikeInfo();
}

export function getPendingAcceptanceRunId(): Promise<unknown> {
  return nativeModule.getPendingAcceptanceRunId();
}

export function runFreshAcceptance(): Promise<unknown> {
  return nativeModule.runFreshAcceptance();
}

export function runResumeAcceptance(runId: string): Promise<unknown> {
  return nativeModule.runResumeAcceptance(runId);
}

export function runNegativeAcceptance(runId: string): Promise<unknown> {
  return nativeModule.runNegativeAcceptance(runId);
}

export function runPerformanceAcceptance(runId: string): Promise<unknown> {
  return nativeModule.runPerformanceAcceptance(runId);
}

export function cleanupAcceptance(runId: string): Promise<unknown> {
  return nativeModule.cleanupAcceptance(runId);
}

export function inspectProductionIdentity(): Promise<unknown> {
  return nativeModule.inspectProductionIdentity();
}

export function prepareIdentityRegistration(displayName: string | null): Promise<unknown> {
  return nativeModule.prepareIdentityRegistration(displayName);
}

export function preparePairingPacket(
  requestId: string,
  expiresAtMs: number,
  targetBundleJson: string,
): Promise<unknown> {
  return nativeModule.preparePairingPacket(requestId, expiresAtMs, targetBundleJson);
}

export function commitIdentityRegistration(
  operationId: string,
  receiptId: string,
): Promise<unknown> {
  return nativeModule.commitIdentityRegistration(operationId, receiptId);
}

export function signDeviceRequest(canonicalRequest: string): Promise<unknown> {
  return nativeModule.signDeviceRequest(canonicalRequest);
}

export function resetProductionIdentity(): Promise<unknown> {
  return nativeModule.resetProductionIdentity();
}

export function respondToPairingRequest(requestId: string, action: string): Promise<unknown> {
  return nativeModule.respondToPairingRequest(requestId, action);
}

export function confirmPairingSafetyNumber(requestId: string): Promise<unknown> {
  return nativeModule.confirmPairingSafetyNumber(requestId);
}

export function activatePairedRelationship(requestId: string, pairId: string): Promise<unknown> {
  return nativeModule.activatePairedRelationship(requestId, pairId);
}

export function listPairingOutbox(): Promise<unknown> {
  return nativeModule.listPairingOutbox();
}

export function ackPairingOutbox(operationId: string, receiptId: string): Promise<unknown> {
  return nativeModule.ackPairingOutbox(operationId, receiptId);
}

export function sweepPairingState(): Promise<unknown> {
  return nativeModule.sweepPairingState();
}

File: modules/m2y-crypto/android/src/main/java/com/m2y/crypto/M2YCryptoModule.kt
package com.m2y.crypto

import android.os.Build
import com.m2y.crypto.production.ProductionIdentityException
import com.m2y.crypto.production.ProductionIdentityManager
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

private const val LIBSIGNAL_VERSION = "0.101.0"
private const val PROTOCOL_ID = "signal-pqxdh-double-ratchet"

class M2YCryptoModule : Module() {
  private val productionManager: ProductionIdentityManager by lazy {
    ProductionIdentityManager(
      appContext.reactContext ?: throw Exceptions.AppContextLost(),
    )
  }

  override fun definition() =
    ModuleDefinition {
      Name("M2YCrypto")

      Function("getSpikeInfo") {
        val nativeLoadVerified = verifyNativeLoad()

        mapOf(
          "abi" to (Build.SUPPORTED_ABIS.firstOrNull() ?: "unknown"),
          "libraryVersion" to LIBSIGNAL_VERSION,
          "nativeLoadVerified" to nativeLoadVerified,
          "platform" to "android",
          "protocol" to PROTOCOL_ID,
        )
      }

      AsyncFunction("runFreshAcceptance") {
        try {
          acceptanceHarness().runFresh()
        } catch (_: LinkageError) {
          throw LibsignalProtocolException()
        } catch (_: Exception) {
          throw LibsignalProtocolException()
        }
      }

      AsyncFunction("getPendingAcceptanceRunId") {
        try {
          acceptanceHarness().findPendingRunId()
        } catch (_: Exception) {
          throw LibsignalProtocolException()
        }
      }

      AsyncFunction("runResumeAcceptance") { runId: String ->
        runAcceptance { it.runResume(runId) }
      }

      AsyncFunction("runNegativeAcceptance") { runId: String ->
        runAcceptance { it.runNegative(runId) }
      }

      AsyncFunction("runPerformanceAcceptance") { runId: String ->
        runAcceptance { it.runPerformance(runId) }
      }

      AsyncFunction("cleanupAcceptance") { runId: String ->
        runAcceptance { it.cleanup(runId) }
      }

      AsyncFunction("inspectProductionIdentity") {
        runProduction { it.inspectProductionIdentity() }
      }

      AsyncFunction("prepareIdentityRegistration") { displayName: String? ->
        runProduction { it.prepareIdentityRegistration(displayName) }
      }

      AsyncFunction("preparePairingPacket") {
          requestId: String,
          expiresAtMs: Long,
          targetBundleJson: String,
        ->
        runProduction { it.preparePairingPacket(requestId, expiresAtMs, targetBundleJson) }
      }

      AsyncFunction("commitIdentityRegistration") { operationId: String, receiptId: String ->
        runProduction { it.commitIdentityRegistration(operationId, receiptId) }
      }

      AsyncFunction("signDeviceRequest") { canonicalRequest: String ->
        runProduction { it.signDeviceRequest(canonicalRequest) }
      }

      AsyncFunction("respondToPairingRequest") { requestId: String, action: String ->
        runProduction { it.respondToPairingRequest(requestId, action) }
      }

      AsyncFunction("confirmPairingSafetyNumber") { requestId: String ->
        runProduction { it.confirmPairingSafetyNumber(requestId) }
      }

      AsyncFunction("activatePairedRelationship") { requestId: String, pairId: String ->
        runProduction { it.activatePairedRelationship(requestId, pairId) }
      }

      AsyncFunction("listPairingOutbox") {
        runProduction { it.listPairingOutbox() }
      }

      AsyncFunction("ackPairingOutbox") { operationId: String, receiptId: String ->
        runProduction { it.ackPairingOutbox(operationId, receiptId) }
      }

      AsyncFunction("sweepPairingState") {
        runProduction { it.sweepPairingState() }
      }

      AsyncFunction("resetProductionIdentity") {
        try {
          productionIdentityManager().resetProductionIdentity()
          mapOf("schemaVersion" to 1, "status" to "reset")
        } catch (error: ProductionIdentityException) {
          throw ProductionIdentityCodedException(error.safeCode())
        }
      }
    }

  private fun acceptanceHarness(): M2YCryptoAcceptanceHarness =
    M2YCryptoAcceptanceHarness(
      appContext.reactContext ?: throw Exceptions.AppContextLost(),
    )

  private fun productionIdentityManager(): ProductionIdentityManager = productionManager

  private fun runProduction(
    operation: (ProductionIdentityManager) -> Map<String, Any>,
  ): Map<String, Any> =
    try {
      operation(productionIdentityManager())
    } catch (error: ProductionIdentityException) {
      throw ProductionIdentityCodedException(error.safeCode())
    } catch (_: LinkageError) {
      throw ProductionIdentityCodedException("identity-native-runtime-unavailable")
    } catch (_: Exception) {
      throw ProductionIdentityCodedException("identity-operation-failed")
    }

  private fun runAcceptance(
    operation: (M2YCryptoAcceptanceHarness) -> Map<String, Any>,
  ): Map<String, Any> =
    try {
      operation(acceptanceHarness())
    } catch (_: LinkageError) {
      throw LibsignalProtocolException()
    } catch (_: Exception) {
      throw LibsignalProtocolException()
    }

  private fun verifyNativeLoad(): Boolean =
    try {
      if (!LibsignalLoadProbe.verify()) {
        throw LibsignalNativeLoadException()
      }
      true
    } catch (_: LinkageError) {
      throw LibsignalNativeLoadException()
    } catch (_: RuntimeException) {
      throw LibsignalNativeLoadException()
    }
}

private class LibsignalNativeLoadException :
  CodedException(
    "E_M2Y_CRYPTO_NATIVE_LOAD",
    "Native cryptography runtime is unavailable.",
    null,
  )

private class LibsignalProtocolException :
  CodedException(
    "E_M2Y_CRYPTO_PROTOCOL",
    "Native protocol acceptance failed.",
    null,
  )

private class ProductionIdentityCodedException(safeCode: String) :
  CodedException(
    "E_M2Y_PRODUCTION_IDENTITY",
    safeCode,
    null,
  )
